core/src/main/scala/filodb.core/memstore/ratelimit/RocksDbCardinalityStore.scala

```diff
@@ -144,108 +160,129 @@ class RocksDbCardinalityStore(ref: DatasetRef, shard: Int) extends CardinalitySt
+    db.put(key, cardinalityNodeToBytes(CardinalityNode.fromCardinality(card)))
   }
 
   def getOrZero(shardKeyPrefix: Seq[String], zero: Cardinality): Cardinality = {
-    val value = db.get(toStringKey(shardKeyPrefix, false).getBytes(StandardCharsets.UTF_8))
-    if (value == NotFound) zero else bytesToCardinality(value)
+    val value = db.get(toStringKey(shardKeyPrefix).getBytes(StandardCharsets.UTF_8))
+    if (value == NotFound) zero else CardinalityNode.toCardinality(bytesToCardinalityNode(value), shardKeyPrefix)
   }
 
   override def remove(shardKeyPrefix: Seq[String]): Unit = {
-    db.delete(toStringKey(shardKeyPrefix, false).getBytes(StandardCharsets.UTF_8))
+    db.delete(toStringKey(shardKeyPrefix).getBytes(StandardCharsets.UTF_8))
   }
 
-  override def scanChildren(shardKeyPrefix: Seq[String]): Seq[Cardinality] = {
-    val it = db.newIterator()
-    val buf = ArrayBuffer[Cardinality]()
-    try {
-      val searchPrefix = toStringKey(shardKeyPrefix, true)
-      logger.debug(s"Scanning shard=$shard dataset=$ref ${new String(searchPrefix)}")
-      it.seek(searchPrefix.getBytes(StandardCharsets.UTF_8))
-      import scala.util.control.Breaks._
-
-      breakable {
-        while (it.isValid()) {
-          val key = new String(it.key(), StandardCharsets.UTF_8)
-          if (key.startsWith(searchPrefix)) {
-            buf += bytesToCardinality(it.value())
-          } else break // dont continue beyond valid results
-          it.next()
+  override def scanChildren(shardKeyPrefix: Seq[String], depth: Int): CardIter = {
+    require(depth > shardKeyPrefix.size,
+      s"scan depth $depth must be greater than the size of the prefix ${shardKeyPrefix.size}")
+
+    new CardIter {
+      val it_ = db.newIterator()
+      var nextCard_ = Cardinality(Seq("should be overwritten before exposed"), -1, -1, -1, -1)
+      val strPrefix_ = toStringKeyPrefix(shardKeyPrefix, depth)
+      var isSeeked = false
+
+      override def seek(): Unit = {
+        logger.debug(s"Scanning shard=$shard dataset=$ref ${new String(strPrefix_)}")
+        it_.seek(strPrefix_.getBytes(StandardCharsets.UTF_8))
+        isSeeked = true
+      }
+
+      // note: must be called exactly once before each next() call
+      override def hasNext: Boolean = {
+        assert(isSeeked, "must call seek() before use")
+        if (it_.isValid) {
+          // store the next matching key and increment the iterator
+          val key = new String(it_.key(), StandardCharsets.UTF_8)
+          if (key.startsWith(strPrefix_)) {
+            val node = bytesToCardinalityNode(it_.value())
+            val prefix = key.split(KeySeparator).drop(1)
+            nextCard_ = CardinalityNode.toCardinality(node, prefix)
+            it_.next()
+            return true
+          }
         }
+        it_.close()
+        false
       }
-    } finally {
-      it.close();
+
+      override def next(): Cardinality = nextCard_
+
+      override def close(): Unit = it_.close()
```

Comment:
Does it need an "if-not-already-closed" check?

Reply:
Nope. From RocksDB:
```
  @Override
  public void close() {
    if (owningHandle_.compareAndSet(true, false)) {
      disposeInternal();
    }
  }
```
But this is not explicitly documented as "`close` is guaranteed idempotent". Want me to add a check anyways (if we keep the iterator)?

Reply:
Irrelevant now-- the iterator is replaced by a Seq